ttl: validate ttl_expiration_expression is a TIMESTAMPTZ

rafiss opened the pull request:
fixes https://github.com/cockroachdb/cockroach/issues/86410
fixes https://github.com/cockroachdb/cockroach/issues/85531
fixes https://github.com/cockroachdb/cockroach/issues/84725

Previously the only validation is that the expression is valid SQL
syntax. Now the expression must also be a TIMESTAMPTZ which will
prevent errors in the TTL job when comparing to the cutoff.

Validation is also added so that a column cannot be dropped or altered
in type if it is used in the TTL expression. If the column is renamed, the
expression is automatically updated.

Release note: None

Release justification: low risk changes to validation

ecwall committed:
ttl: validate ttl_expiration_expression is a TIMESTAMPTZ

Previously the only validation is that the expression is valid SQL
syntax. Now the expression must also be a TIMESTAMPTZ which will
prevent errors in the TTL job when comparing to the cutoff.

Validation is also added so that a column cannot be dropped or altered
in type if it is used in the TTL expression. If the column is renamed, the
expression is automatically updated.

Release note: None

Release justification: high priority fix to new functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86269)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
Nice, thanks for doing this!

> Reviewed 1 of 11 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-:-N9cu2Uw2T7CzOQ_4CGY:bx668bm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[`pkg/sql/alter_table.go` line 497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9cuBW-C60yqBgtqaQv:-N9cuBW-C60yqBgtqaQw:bqd8xk5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/alter_table.go#L497)):*
> ```Go
> 							t.Column,
> 						),
> 						"use ALTER TABLE %s RESET (ttl) or ALTER TABLE %s SET (ttl_expiration_expression = ...) instead",
> ```

need to repeat the argument or use a positional format specifier 
___
*[`pkg/sql/catalog/schemaexpr/expr.go` line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9csingEkVwkzIDO6k0:-N9csingEkVwkzIDO6k1:b-owkhhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/catalog/schemaexpr/expr.go#L461)):*
> ```Go
> // ValidateTTLExpirationExpressionHasNoDependents verifies that the
> // ttl_expiration_expression, if any, does not reference the given column.
> func ValidateTTLExpirationExpressionHasNoDependents(
> ```

this is a weird name given you're validating that the `TTLExpression` does not depend on the column. Maybe this should be `ValidateTTLExpressionDoesNotDependOnColumn`? 
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ct4IH3wJiapPACOuU:-N9ct4IH3wJiapPACOuV:bs9vxym) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/catalog/tabledesc/validate.go#L786)):*
> ```Go
> 	})
> 
> 	vea.Report(ValidateRowLevelTTL(desc.GetRowLevelTTL()))
> ```

Any reason not to add `desc` as the argument to `ValidateRowLevelTTL` and call this or add the logic there? It looked to me like all the callers have a descriptor handy.



_Code quote:_
```Go
	vea.Report(ValidateRowLevelTTL(desc.GetRowLevelTTL()))
	if desc.HasRowLevelTTL() {
```
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctcAh71hzPky5N-2X:-N9ctcAh71hzPky5N-2Y:b-tslfa5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/builder_state.go#L473)):*
> ```Go
> 		panic(errors.AssertionFailedf("failed to get descriptor for table %d", tableID))
> 	}
> 	tableDesc, ok := cachedDesc.desc.(catalog.TableDescriptor)
> ```

can you revise this error to say that it's not a table and print the type with `%T` or `DescriptorType()`



_Code quote:_
```Go
	tableDesc, ok := cachedDesc.desc.(catalog.TableDescriptor)
	if !ok {
		panic(errors.AssertionFailedf("failed to get descriptor for table %d", tableID))
```
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctVPe0qZ4Ow1-2g6c:-N9ctVPe0qZ4Ow1-2g6d:b-5mfh7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L78)):*
> ```Go
> 	})
> 
> 	if rowLevelTTL != nil {
> ```

nit: invert the condition and early return?
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctK3k99H76rF22W6C:-N9ctK3k99H76rF22W6D:b-ltenze) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L97)):*
> ```Go
> 			}
> 			wrappedExpr := b.WrapExpression(tbl.TableID, expr)
> 			for _, referencedCol := range wrappedExpr.ReferencedColumnIDs {
> ```

`if descpb.ColumnIDs(wrappedExpr.ReferencedColumnIDs).Contains(colToDrop.ColumnID) {`?



_Code quote:_
```Go
			for _, referencedCol := range wrappedExpr.ReferencedColumnIDs {
				if referencedCol == colToDrop.ColumnID {
```


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86269)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @ecwall)

___
*[`pkg/sql/alter_table.go` line 497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9cuBW-C60yqBgtqaQv:-N9hrjWSEpPhAvbasges:b-qssb2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/alter_table.go#L497)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

need to repeat the argument or use a positional format specifier 
</blockquote></details>

fixing
___
*[`pkg/sql/catalog/schemaexpr/expr.go` line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9csingEkVwkzIDO6k0:-N9hrkxN0hsxKzhIJv7c:b7anmnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/catalog/schemaexpr/expr.go#L461)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

this is a weird name given you're validating that the `TTLExpression` does not depend on the column. Maybe this should be `ValidateTTLExpressionDoesNotDependOnColumn`? 
</blockquote></details>

good call
___
*[`pkg/sql/catalog/tabledesc/validate.go` line 786 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ct4IH3wJiapPACOuU:-N9hrlzkAn1Efpnc7RR4:bvukauk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/catalog/tabledesc/validate.go#L786)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Any reason not to add `desc` as the argument to `ValidateRowLevelTTL` and call this or add the logic there? It looked to me like all the callers have a descriptor handy.
</blockquote></details>

i attempted to explain in the comment below. the reason is that the usage at https://github.com/cockroachdb/cockroach/blob/79ef82000093190c12991edda480866154536b85/pkg/sql/storageparam/tablestorageparam/table_storage_param.go#L49 can be called with a relatively empty table descriptor when the `ttl_expiration_expression` param is configured as part of `CREATE TABLE`
___
*[`pkg/sql/schemachanger/scbuild/builder_state.go` line 473 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctcAh71hzPky5N-2X:-N9hs9uk2eDSvFMzwzqx:b-qssb2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/builder_state.go#L473)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you revise this error to say that it's not a table and print the type with `%T` or `DescriptorType()`
</blockquote></details>

fixing
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctVPe0qZ4Ow1-2g6c:-N9hsAiW-doEyrYJN2fX:b-qssb2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L78)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: invert the condition and early return?
</blockquote></details>

fixing
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 97 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9ctK3k99H76rF22W6C:-N9hsCK1DjEOxRBVUuPY:b-qssb2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/d663614033eb4f39755402ed07b1200e596fc569/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L97)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

`if descpb.ColumnIDs(wrappedExpr.ReferencedColumnIDs).Contains(colToDrop.ColumnID) {`?
</blockquote></details>

fixing


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

just needs a rewrite of `pkg/sql/schemachanger/scbuild`

> Reviewed 5 of 11 files at r1, 5 of 8 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-:-N9nSgwU2XX222aJhNIa:b-bfupa4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @ecwall, and @rafiss)

___
*[`pkg/sql/alter_table.go` line 1997 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86269#-N9nS8fY4x2vpB6vgPIN:-N9nS8fY4x2vpB6vgPIO:b-rtxi4v) ([raw file](https://github.com/cockroachdb/cockroach/blob/2747b6fbea68cf2f910d973ce767c68843f6d252/pkg/sql/alter_table.go#L1997)):*
> ```Go
> 	}
> 
> 	// validate ttl_expiration_expression
> ```

nit: fix up this comment.


<!-- Sent from Reviewable.io -->

postamar commented:
I appreciate that this PR includes significant coverage in the form of logic tests, still, this seems like a good place to add all of the missing unit test cases in `tabledesc` for the whole TTL configuration in the table descriptors, as tracked in https://github.com/cockroachdb/cockroach/issues/86410

rafiss committed:
catalog/tabledesc: add tests for TTL validation and cleanup

This moves more of the TTL validation into ttl.go, and outside of the
CrossTableReferences validation.

Tests are added for all the TTL validation steps.

Release justification: test only change and small refactor
Release note: None

rafiss commented:
@postamar i've added a separate commit that focuses on TTL validation testing.

postamar reviewed (approved):
Thanks! LGTM

rafiss commented:
TFTRs!

bors r=ajwerner,postamar

rafiss commented:
bors r=ajwerner,postamar
